### PR TITLE
feat(smart-contract-orders): hide presign order until wallet returns tx for Safe app limit bundling tx

### PR DESCRIPTION
**Pull request by alfetopito**
# Summary

Hide order from UI until it's confirmed by the wallet

https://github.com/cowprotocol/cowswap/assets/43217/78224eed-8042-47b8-8df0-56a00a29a325

If user rejects the tx or it's never signed, the order will remain hidden in the same device.
It'll still be visible in different devices and on the Explorer

## ⚠️ Notes ⚠️
- For now the **only** path implemented is for Safe wallet, loaded as an App, for Limit orders, bundling approve + presign
- Will implement remaining paths/wallet in a follow up PR
- This is an experimental PR and I'm focusing on only one flow as the major changes in the state/middleware

  # To Test

1. Load app inside the Safe and open the limit orders
2. Pick a pair of which the sell token needs approval
3. Confirm the order until the Safe confirmation is displayed
* By this point, the order is already created on the backend and normally you would see the pop up and hear the sound
* There should be no pop up and no sound
4. Reject the tx
* Before this change, you would see the order in the list in a pre-sign stage
* The new order should not be visible anywhere
5. Open the Explorer and check the orders for you account
* The order should be shown there as presign pending state
6. Repeat step `3`, but this time submit/execute the tx
* The order should now appear is presign state
* There should be a pop up
* The moo sound should play
* Once executed/mined, the order should change to open and follow the regular flow

Please also test the EOA and Ethflow notifications and sounds, as this change touches on all of these areas.
Everything should behave the same as before.

### CODE REVIEW

File: src/state/orders/middleware.tsx [hunk 1]
```diff
@@ -58,46 +56,51 @@ export const popupMiddleware: Middleware<Record<string, unknown>, AppState> = (s
...
-    } else if (isCancelOrderAction(action)) {
-      // action is order/cancelOrder
-      // Cancelled Order Popup
-      popup = _buildCancellationPopup(orderObject.order)
-      orderAnalytics('Canceled', orderClass)
-    } else {
-      // action is order/expireOrder
-      // Expired Order Popup
-      popup = setPopupData(OrderTxTypes.METATXN, {
-        success: false,
```
Comment (alfetopito) [inline src/state/orders/middleware.tsx:89]: These actions were never reached as `isSingleOrderChangeAction` only contained a single action: `isPendingOrderAction`

**Comment by elena-zh** [comment]
Hey @alfetopito , cool! Again, thank you for the perfect PR description!

I have run tests, and here are my notes: 
1. (should not be related to the current PR, however, might be fixed in the upcoming PRs): once we specified but not signed 'Approve+Swap' market order (not limit), there is a warning in the Your orders table saying that it is a **limit** order. Actually, it is a market order. Can we remove an order type from the warning in order to make it a bit more generic?
market

2. I still can see 'Signing' limit orders when I disconnect/connect to the wallet (and refresh the page). I assume that this is happening due to orders are loaded from API, but I'd better clarify it with you if this is OK or not :)
refresh the page

3. Can we deal somehow with one more case: I sign an order, then reject the order transaction in the Safe (and execute rejection). 
regected

Thank you!

**Comment by alfetopito** [comment]
> Hey @alfetopito , cool! Again, thank you for the perfect PR description!
> 
> I have run tests, and here are my notes:
> 
>     1. (should not be related to the current PR, however, might be fixed in the upcoming PRs): once we specified but not signed 'Approve+Swap' market order (not limit), there is a warning in the Your orders table saying that it is a **limit** order. Actually, it is a market order. Can we remove an order type from the warning in order to make it a bit more generic?
>        market
> 
>     2. I still can see 'Signing' limit orders when I disconnect/connect to the wallet (and refresh the page). I assume that this is happening due to orders are loaded from API, but I'd better clarify it with you if this is OK or not :)
>        refresh the page
> 
>     3. Can we deal somehow with one more case: I sign an order, then reject the order transaction in the Safe (and execute rejection).
>        regected
> 
> 
> Thank you!

**Comment by alfetopito** [comment]
Closed by accident

**Comment by elena-zh** [comment]
>     2. I still can see 'Signing' limit orders when I disconnect/connect to the wallet (and refresh the page). I assume that this is happening due to orders are loaded from API, but I'd better clarify it with you if this is OK or not :)
>        refresh the page

Is it the same browser? I'll take a look, but the orders should remain hidden

Yes, the same. 
I sent an order to sign, did not sign. Then I refreshed the page, and it appeared that my MM locked and was disconnected. 
I reconnected Safe to MM once again, and noticed this 'Signing' order.

**Review by elena-zh — APPROVED** [review]

File: src/state/enhancedTransactions/updater/FinalizeTxUpdater.tsx [hunk 2]
```diff
@@ -12,16 +12,17 @@ import { checkedTransaction, finalizeTransaction, updateSafeTransaction } from '
...
 import { GetSafeInfo, useGetSafeInfo } from 'hooks/useGetSafeInfo'
 import { useWeb3React } from '@web3-react/core'
 import { supportedChainId } from 'utils/supportedChainId'
-import { cancelOrdersBatch, invalidateOrdersBatch, updateOrder } from 'state/orders/actions'
+import { cancelOrdersBatch, invalidateOrdersBatch } from 'state/orders/actions'
 import { useSetAtom } from 'jotai'
 import { removeInFlightOrderIdAtom } from '@cow/modules/swap/state/EthFlow/ethFlowInFlightOrderIdsAtom'
 import ms from 'ms.macro'
 import { useWalletInfo } from '@cow/modules/wallet'
+import { partialOrderUpdate } from '@src/state/orders/utils'
```
Comment (shoom3301) [inline src/state/enhancedTransactions/updater/FinalizeTxUpdater.tsx:24]:
```suggestion
import { partialOrderUpdate } from 'state/orders/utils'
```

File: src/state/enhancedTransactions/updater/FinalizeTxUpdater.tsx [hunk 3]
```diff
@@ -12,16 +12,17 @@ import { checkedTransaction, finalizeTransaction, updateSafeTransaction } from '
...
+import { AppDispatch } from '@src/state'
```
Comment (shoom3301) [inline src/state/enhancedTransactions/updater/FinalizeTxUpdater.tsx:25]:
```suggestion
import { AppDispatch } from 'state'
```

**Comment by alfetopito** [comment]
> I saw it locally the case Elena mentioned. Putting it on hold as it need to be addressed before merging it, and for now we want to focus on finishing the MVP

Not sure the expired orders I saw locally were hidden before.
Checking the localStorage orders, I see several with the `isHidden` flag `true`.
Also could not reproduce it again, even though I tried it many different ways.

It could be that I saw them when looking at a different PR/branch which did not have the filtering logic, of which makes sense to show the ones that should be hidden.

cc @elena-zh

File: src/legacy/state/orders/middleware.tsx [hunk 4]
```diff
@@ -58,46 +56,51 @@ export const popupMiddleware: Middleware<Record<string, unknown>, AppState> = (s
...
-      // Cancelled Order Popup
-      popup = _buildCancellationPopup(orderObject.order)
-      orderAnalytics('Canceled', orderClass)
-    } else {
-      // action is order/expireOrder
-      // Expired Order Popup
-      popup = setPopupData(OrderTxTypes.METATXN, {
-        success: false,
+
+    const hash = orderObject.order.orderCreationHash
```
Comment (shoom3301) [inline src/legacy/state/orders/middleware.tsx:60]: Nitpick:
The logic is pretty complex (at least for me). Would be great to cover it by unit-tests

Comment (alfetopito) [reply src/legacy/state/orders/middleware.tsx:60]: https://github.com/cowprotocol/cowswap/pull/2552

**Pull request merged**